x/wasm/types/errors.go

```diff
@@ -164,3 +152,32 @@ func (e WasmVMFlavouredError) Wrap(desc string) error { return errorsmod.Wrap(e,
 func (e WasmVMFlavouredError) Wrapf(desc string, args ...interface{}) error {
 	return errorsmod.Wrapf(e, desc, args...)
 }
+
+// DeterministicError is a wrapper type around an error that the creator guarantees to have
+// a deterministic error message.
+// This means that the `Error()` function must always return the same string on all nodes.
+// DeterministicErrors are not redacted when returned to a contract,
+// so not upholding this guarantee can lead to consensus failures.
+type DeterministicError struct {
+	error
+}
+
+var _ error = DeterministicError{}
+
+// MarkErrorDeterministic marks an error as deterministic.
+// Make sure to only do that if the error message is deterministic between systems.
+// See [DeterministicError] for more details.
+func MarkErrorDeterministic(e error) DeterministicError {
+	return DeterministicError{error: e}
+}
+
+// Unwrap implements the built-in errors.Unwrap
+func (e DeterministicError) Unwrap() error {
+	return e.error
+}
+
+// Cause is the same as unwrap but used by ABCIInfo
```

Comment:
Is there an interface we can check this implementation against?

Reply:
Unfortunately not. ABCIInfo has a [private interface](https://github.com/cosmos/cosmos-sdk/blob/35c4243da9c8c285a4b65117238dec9b9f7921c8/errors/errors.go#L300) for that and uses it to [iterate through the error chain](https://github.com/cosmos/cosmos-sdk/blob/35c4243da9c8c285a4b65117238dec9b9f7921c8/errors/abci.go#L92) until it finds an error with codespace / code.

Reply:
okay, fine, thanks for checking